CMakeLists.txt

```
@@ -63,8 +63,21 @@ configure_file("${CMAKE_SOURCE_DIR}/cmake/templates/license.h.in" include/licens
 
 include(EthOptions)
 configure_project(TESTS)
-
+set(LATEST_Z3_VERSION "4.8.12")
 find_package(Z3 4.8.0)
+# NOTE: With EMSCRIPTEN=ON we do not build code from test/ dir and that
+# code is our only reason to require latest Z3.
+if (TESTS AND NOT EMSCRIPTEN)
+  if (${Z3_FOUND})
+    if ( NOT ( "${Z3_VERSION_STRING}" VERSION_EQUAL ${LATEST_Z3_VERSION} ) )
+      message(
+        FATAL_ERROR
+        "SMTChecker tests require Z3 ${LATEST_Z3_VERSION} for all tests to pass.\n\
+Build with -DTESTS=OFF if you want to use an older version. \
+You can also use -DUSE_Z3=OFF to build without Z3 and then use --no-smt when running tests.")
```

So going with https://github.com/ethereum/solidity/pull/11873#discussion_r706342315, we need to change it to:
```suggestion
        "SMTChecker tests require Z3 ${LATEST_Z3_VERSION} for all tests to pass.\n\
If you want to build without Z3, pass -DUSE_Z3=OFF to cmake and later use --no-smt when running tests."
      )
```

---

Oh, this should also mention the `-DSTRICT_Z3_VERSION` flag. Suggestion updated.

---

done